src/observer.ts

```diff
@@ -34,7 +34,7 @@ export function observer<
                   >
               >
         : never /* forwardRef set for a non forwarding component */
-    : C & React.FunctionComponent
+    : C extends React.FunctionComponent<infer P> ? C & React.FunctionComponent<P> : never /* observer uses for non functional component */
```

Comment:
I think this should just read `C`, which is already a React.FunctionComponent<P> per line 23, so need to infer and repeat the props. I'm pretty sure the definition on line 17 should be moved below this one, as the typings of this function are more specific. On line 30 I don't think the `C &` can be dropped as well.

Reply:
I don't quite understand why all these definitions if one definition can handle both variants.
```ts
export declare function observer<C extends React.FunctionComponent<any> | React.RefForwardingComponent<any, any>, Options extends IObserverOptions>(baseComponent: C, options?: Options): Options extends { forwardRef: true; }
    ? C extends React.RefForwardingComponent<infer TRef, infer P> ? 
        React.MemoExoticComponent<React.ForwardRefExoticComponent<React.PropsWithoutRef<P> & React.RefAttributes<TRef>>> : never
    : C;
```